Validity and timelocks

paweljakubas opened the pull request:
<!--
Detail in a few bullet points the work accomplished in this PR.

Before you submit, don't forget to:

* Make sure the GitHub PR fields are correct:
   ✓ Set a good Title for your PR.
   ✓ Assign yourself to the PR.
   ✓ Assign one or more reviewer(s).
   ✓ Link to a Jira issue, and/or other GitHub issues or PRs.
   ✓ In the PR description delete any empty sections
     and all text commented in <!--, so that this text does not appear
     in merge commit messages.

* Don't waste reviewers' time:
   ✓ If it's a draft, select the Create Draft PR option.
   ✓ Self-review your changes to make sure nothing unexpected slipped through.

* Try to make your intent clear:
   ✓ Write a good Description that explains what this PR is meant to do.
   ✓ Jira will detect and link to this PR once created, but you can also
     link this PR in the description of the corresponding Jira ticket.
   ✓ Highlight what Testing you have done.
   ✓ Acknowledge any changes required to the Documentation.
-->


- [x] reading properly validity from request body
- [x] add checks on boundary precedence and quantity
- [x] simplify how asset quantity is called in constructTransaction (in line with decodeTransaction)
- [x] slot check
- [x] make either validity bound optional
- [x] introduce before bound to transaction context and propagate this to node's tx
- [x] extend decodeTransaction to detect txValidityInterval (with integration testing)
- [x] introducing slot computation of scripts
- [x] script vs validity check
- [x] timelock minting test on   
- [x] add unit tests in MintBurnSpec

### Comments

<!-- Additional comments, links, or screenshots to attach, if any. -->

### Issue Number
adp-1193

<!-- Reference the Jira/GitHub issue that this PR relates to, and which requirements it tackles.
  Note: Jira issues of the form ADP- will be auto-linked. -->

paweljakubas committed:
add proper usage of validity interval - part 1

paweljakubas committed:
handle invalid validity bounds

paweljakubas committed:
add next check

paweljakubas committed:
simplify how asset quantity is called during construction

paweljakubas committed:
deal with negative slots

paweljakubas committed:
make validity bound optional

paweljakubas committed:
some improvements

paweljakubas committed:
fist timelock minting working

paweljakubas committed:
introduce lower bound to transaction context

paweljakubas committed:
refine integration test

revert flock

paweljakubas committed:
add support of validity in decodeTx - part 1

paweljakubas committed:
add support of validity in decodeTx - part 2

add Data.Interval

add Data.IntervalSpec

revert adding Data.Interval

paweljakubas committed:
adding slot interval computation in scripts

Commit:
Regenerate nix

paweljakubas committed:
compare tx validity with timelocks

paweljakubas committed:
add integration test showing the interval/timelock case

paweljakubas committed:
add toSlotInterval unit tests

paweljakubas committed:
add unit testing for withinSlotInterval

Commit:
Regenerate nix

jonathanknowles committed:
Formatting fixes.

- use 4-space indents.
- avoid variable-length indentation.
- limit line lengths to 80 characters where practical.
- use only a single blank line between top-level definitions.

jonathanknowles committed:
Simplify definition of `fromValidityBound`.

We use `LambdaCase` to remove repetition of the function name, and
factor out the common call to `liftIO`.

jonathanknowles committed:
Use list append in `toSlotInterval` instead of concatenation.

jonathanknowles committed:
Use `LambdaCase` to simplify definitions within `toSlotInterval`.

jonathanknowles committed:
Use eta-reduction to simplify `withinSlotInterval`.

jonathanknowles committed:
Revise API validity interval descriptions and error messages.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -4084,8 +4087,8 @@ data ApiMintData (n :: NetworkDiscriminant) = ApiMintData
-    , amount
-        :: Quantity "assets" Natural
+    , quantity
+        :: Natural
```
Hi @paweljakubas. Is this a breaking change to the API?

paweljakubas replied:
well, yes. but new workflow is still experimental API ;-)

jonathanknowles committed:
Replace `fromIntegral` with safer `intCast` and `Finite` constructor.

Justification:

- Using `fromIntegral` can hide unsafe integral narrowing conversions.
- Using `intCast` eliminates this class of error statically.

When creating values of `Data.Interval.Extended`, we can use the
`Finite` constructor instead of `fromIntegral`.

The following expressions are equivalent:

> Finite 1000 :: Extended Int
Finite 1000

> fromIntegral 1000 :: Extended Int
Finite 1000

> Finite 1000 == fromIntegral 1000
True

jonathanknowles committed:
Rename `slotInterval` to `scriptSlotIntervals`.

This function takes a script and returns a list of slot intervals, rather
than just a single interval.

paweljakubas committed:
scriptSlotIntervals enhancement

piotr-iohk commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/TransactionsNew.hs:
```diff
@@ -1232,6 +1263,58 @@ spec = describe "NEW_SHELLEY_TRANSACTIONS" $ do
+                { "withdrawal": "self"
+                , "validity_interval":
+                    { "invalid_before": "unspecified"
+                    , "invalid_hereafter": "unspecified"
```
The "unspecified" should be removed from the swagger then:
![Screenshot from 2022-04-29 10-03-19](https://user-images.githubusercontent.com/42900201/165906590-4d76c657-17b4-42bd-b5ac-154233b4975d.png)

paweljakubas committed:
integration code reshuffling to soothe plutus

Commit:
remove unspecified from swagger

Commit:
Addtional comment to validity_interval
